src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -6422,15 +6434,33 @@ public object VisitIndexExpression(IndexExpressionAst indexExpressionAst)
             // In the former case, the user is requesting an array slice.  In the latter case, they index expression is likely
             // an array (dynamically determined) and they don't want an array slice, they want to use the array as the index
             // expression.
-            if (arrayLiteral != null && arrayLiteral.Elements.Count > 1)
-            {
-                return DynamicExpression.Dynamic(
-                                            PSGetIndexBinder.Get(arrayLiteral.Elements.Count, constraints),
-                                            typeof(object),
-                                            arrayLiteral.Elements.Select(CompileExpressionOperand).Prepend(targetExpr));
-            }
+            Expression indexingExpr = arrayLiteral != null && arrayLiteral.Elements.Count > 1
+                ? DynamicExpression.Dynamic(
+                    PSGetIndexBinder.Get(arrayLiteral.Elements.Count, constraints),
+                    typeof(object),
+                    arrayLiteral.Elements.Select(CompileExpressionOperand).Prepend(targetExpr))
+                : DynamicExpression.Dynamic(
+                    PSGetIndexBinder.Get(1, constraints),
+                    typeof(object),
+                    targetExpr,
+                    CompileExpressionOperand(index));
+
+            return indexExpressionAst.NullConditional ? GetNullConditionalWrappedExpression(targetExpr, indexingExpr) : indexingExpr;
+        }
+
+        private static Expression GetNullConditionalWrappedExpression(Expression targetExpr, Expression memberAccessExpression)
+        {
+            ParameterExpression varExpr = Expression.Variable(typeof(object));
 
-            return DynamicExpression.Dynamic(PSGetIndexBinder.Get(1, constraints), typeof(object), targetExpr, CompileExpressionOperand(index));
+            ConditionalExpression nullCheckExpression = Expression.IfThenElse(
+                Expression.Call(CachedReflectionInfo.LanguagePrimitives_IsNullLike, targetExpr.Cast(typeof(object))),
+                Expression.Assign(varExpr, ExpressionCache.NullConstant),
+                Expression.Assign(varExpr, memberAccessExpression.Cast(typeof(object))));
+
+            return Expression.Block(
```

Comment:
What's the reason we need a block with an intermediate variable rather than just

```csharp
return Expression.IfThenElse(
    Expression.Call(
        CachedReflectionInfo.LanguagePrimitives_IsNullLike,
        targetExpr.Cast(typeof(object))),
    ExpressionCache.NullConstant,
    memberAccessExpression.Cast(typeof(object)));
```

Reply:
The `IfThenElse` has a return type of void. So nothing is returned back. Hence, we need to save the value to a variable and then return it.

Reply:
@adityapatwardhan you can use `Expression.Condition` instead.  It's a `IfThenElse` that must return.

Reply:
Oh, yeah that's the one I'm thinking of!

Reply:
Changed to use `Expression.Condition` instead.